test/e2e/kubernetes/kubernetes_test.go

```diff
@@ -2008,6 +2024,111 @@ var _ = Describe("Azure Container Cluster using the Kubernetes Orchestrator", fu
 				Skip("No windows agent was provisioned for this Cluster Definition")
 			}
 		})
+		// This test is not parallelizable due to tainting nodes with NoSchedule
+		It("should expect containers to be recreated after node restart", func() {
+			if eng.HasWindowsAgents() {
+				for _, profile := range eng.ExpandedDefinition.Properties.AgentPoolProfiles {
+					if profile.IsWindows() {
+						if profile.AvailabilityProfile == api.AvailabilitySet {
+							Skip("AvailabilitySet is configured for this Cluster Definition")
+						}
+					}
+				}
+
+				windowsImages, err := eng.GetWindowsTestImages()
+				Expect(err).NotTo(HaveOccurred())
+				r := rand.New(rand.NewSource(time.Now().UnixNano()))
+				deploymentPrefix := fmt.Sprintf("iis-%s", cfg.Name)
+				deploymentName := fmt.Sprintf("%s-%v", deploymentPrefix, r.Intn(99999))
+				By("Creating a deployment with 1 pod running IIS")
+				iisDeploy, err := deployment.CreateWindowsDeployWithHostportDeleteIfExist(deploymentPrefix, windowsImages.IIS, deploymentName, "default", 80, -1)
+				Expect(err).NotTo(HaveOccurred())
+
+				By("Waiting on pod to be Ready")
+				running, err := pod.WaitOnSuccesses(deploymentName, "default", 4, sleepBetweenRetriesWhenWaitingForPodReady, cfg.Timeout)
+				Expect(err).NotTo(HaveOccurred())
+				Expect(running).To(Equal(true))
+
+				By("Exposing a LoadBalancer for the pod")
+				err = iisDeploy.ExposeDeleteIfExist(deploymentPrefix, "default", "LoadBalancer", 80, 80)
+				Expect(err).NotTo(HaveOccurred())
+				iisService, err := service.Get(deploymentName, "default")
+				Expect(err).NotTo(HaveOccurred())
+				err = iisService.WaitForIngress(cfg.LBTimeout, 5*time.Second)
+				Expect(err).NotTo(HaveOccurred())
+
+				By("Verifying that the service is reachable and returns the default IIS start page")
+				err = iisService.ValidateWithRetry("(IIS Windows Server)", sleepBetweenRetriesWhenWaitingForPodReady, cfg.Timeout)
+				Expect(err).NotTo(HaveOccurred())
+
+				pods, err := iisDeploy.Pods()
+				Expect(err).NotTo(HaveOccurred())
+				nodeName := pods[0].Spec.NodeName
+				ctx, cancel := context.WithTimeout(context.Background(), 6000*time.Second)
+				defer cancel()
+
+				By("Adding taint to all other Windows nodes")
+				nodeList, err := node.Get()
+				for _, n := range nodeList.Nodes {
+					if n.IsWindows() && n.Metadata.Name != nodeName {
+						n.AddTaint(node.Taint{Key: "key", Value: "value", Effect: "NoSchedule"})
+					}
+				}
+
+				// Removing taints
+				defer func(nodeList *node.List, nodeName string) {
+					for _, n := range nodeList.Nodes {
+						if n.IsWindows() && n.Metadata.Name != nodeName {
+							n.RemoveTaint(node.Taint{Key: "key", Value: "value", Effect: "NoSchedule"})
+						}
+					}
+				}(nodeList, nodeName)
```

Comment:
Do you know if this defer func fires after just this test case.
My concern here is that if this test fails and we don't un-taint the nodes we could skew the resulys of the rest of the test cases.
If would be good to avoid that if we can.

Reply:
I believe it'll execute after the function in line 2028 finishes

Reply:
I verified from the PR tests that the taints are removed after the test